### PR TITLE
Add Instana exporter overall structure

### PR DESCRIPTION
**Pull request by hickeyma**
Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Cedric Ziel <cedric@cedric-ziel.com>
Co-authored-by: Willian Carvalho <willianc@pm.me>

**Description:** 

Adding new component, Instana exporter. 

This is the first of 3 PRs which adds the overall skeleton structure.

**Link to tracking Issue:** 

Partial #13395 

**Testing:** 

Unit Tests

**Documentation:**

Readme explaining the exporter and its configuration.

### CODE REVIEW

File: exporter/instanaexporter/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# Instana Exporter
+
+| Status                   |                  |
+| ------------------------ |------------------|
+| Stability                | [beta]           |
```
Comment (jpkrohling) [inline exporter/instanaexporter/README.md:5]: Please, review the possible values and select accordingly. For this PR, it should probably be "in development", switching to "alpha" once the component is ready. "beta" implies that this component has had at least "some" exposure to production workloads.

File: exporter/instanaexporter/config/config.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"strings"
+
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/confighttp"
+)
+
+const (
+	// AttributeInstanaHostID can be used to distinguish multiple hosts' data
+	// being processed by a single collector (in a chained scenario)
+	AttributeInstanaHostID = "instana.host.id"
```
Comment (jpkrohling) [inline exporter/instanaexporter/config/config.go:28]: Do you really mean them all to be exported?

File: exporter/instanaexporter/factory.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+
+	instanaConfig "github.com/ibm-observability/instanaexporter/config"
+)
+
+const (
+	// The value of "type" key in configuration.
+	typeStr = "instana"
+	// The stability level of the exporter.
+	stability = component.StabilityLevelBeta
```
Comment (jpkrohling) [inline exporter/instanaexporter/factory.go:34]: In development

File: exporter/instanaexporter/factory.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+// createDefaultConfig creates the default exporter configuration
+func createDefaultConfig() config.Exporter {
+	return &instanaConfig.Config{
+		ExporterSettings: config.NewExporterSettings(config.NewComponentID(typeStr)),
+		HTTPClientSettings: confighttp.HTTPClientSettings{
+			Endpoint: "",
+			Timeout:  30 * time.Second,
+			Headers:  map[string]string{},
+			// We almost read 0 bytes, so no need to tune ReadBufferSize.
```
Comment (jpkrohling) [inline exporter/instanaexporter/factory.go:54]: Not sure I get this comment.

File: exporter/instanaexporter/factory.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	return exporterhelper.NewTracesExporterWithContext(
+		ctx,
+		set,
+		config,
+		instanaExporter.pushConvertedTraces,
```
Comment (jpkrohling) [inline exporter/instanaexporter/factory.go:76]: Where is this one?

**Review by jpkrohling — COMMENTED** [review]
This PR seems to be missing a few things from the skeleton, like the Makefile.

File: exporter/instanaexporter/README.md (hunk 1 above)
Comment (hickeyma) [reply exporter/instanaexporter/README.md]: Agree, updated.

File: exporter/instanaexporter/config/config.go (hunk 2 above)
Comment (hickeyma) [reply exporter/instanaexporter/config/config.go]: Good catch, moved to internal.

File: exporter/instanaexporter/factory.go (hunk 3 above)
Comment (hickeyma) [reply exporter/instanaexporter/factory.go]: Done

File: exporter/instanaexporter/factory.go (hunk 4 above)
Comment (hickeyma) [reply exporter/instanaexporter/factory.go]: Agree, removed

File: exporter/instanaexporter/factory.go (hunk 5 above)
Comment (hickeyma) [reply exporter/instanaexporter/factory.go:75]: It is part of the Instana implementation code. This is pushed in the 2nd PR.

**Comment by hickeyma** [comment]
@jpkrohling Thanks for the review and feedback.

I have updated your comments and it is ready for review again.

> This PR seems to be missing a few things from the skeleton, like the Makefile

I am following the [When adding a new component](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component) and for 1st PR it suggests the: "Readme, configuration, and factory implementation". The Makefile etc. can be added in 2nd PR with the implementation, so it can build.

**Comment by jpkrohling** [comment]
> The Makefile etc. can be added in 2nd PR with the implementation, so it can build.

Interesting, so far, our interpretation has been that the Makefile has to be part of the first PR so that it's incorporated into the main build. This way, we can ensure the files are syntactically correct, like the `config.go`. I believe the only thing you'd have to change in this PR is remove the reference to the non-existing function from the factory.

**Comment by hickeyma** [comment]
Thanks @jpkrohling for the feedback. I have added the Makefile as you suggested. Let me know if it needs any other changes.

File: .github/CODEOWNERS [hunk 6]
```diff
@@ -45,6 +45,7 @@ exporter/googlemanagedprometheusexporter/            @open-telemetry/collector-c
 exporter/googlecloudpubsubexporter/                  @open-telemetry/collector-contrib-approvers @alexvanboxel
 exporter/humioexporter/                              @open-telemetry/collector-contrib-approvers @xitric
 exporter/influxdbexporter/                           @open-telemetry/collector-contrib-approvers @jacobmarble
+exporter/instanaexporter/                            @open-telemetry/collector-contrib-approvers @hickeyma
```
Comment (jpkrohling) [inline .github/CODEOWNERS:48]: Given I'm the sponsor, I believe I should be part of the code owners for this component.

File: exporter/instanaexporter/README.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
+# Instana Exporter
+
+| Status                   |                  |
+| ------------------------ |------------------|
+| Stability                | [in-development]  |
```
Comment (jpkrohling) [inline exporter/instanaexporter/README.md:5]: Not for this PR, but the table deserves properly formatting :-)

**Review by jpkrohling — APPROVED** [review]

File: exporter/instanaexporter/README.md (hunk 7 above)
Comment (hickeyma) [reply exporter/instanaexporter/README.md:5]: Its is unaligned here in the comparison but it is aligned when the markdown is rendered.

File: exporter/instanaexporter/factory.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	"context"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+
+	instanaConfig "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/instanaexporter/config"
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply exporter/instanaexporter/factory.go:27]: Just a throw back when internal constants were in the config. Moved it to the root now.

**Comment by hickeyma** [comment]
@jpkrohling Thanks for the reviews and feedback. I have updated it and it should be good for review again.

**Review by jpkrohling — APPROVED** [review]

**Comment by hickeyma** [comment]
Thanks @jpkrohling for review, feedback and merging.

**Pull request merged**